### ISSUE TITLE
buildifier lack of line length limits leads to awful list comprehensions

### ISSUE THREAD

**Issue by shahms** [bazelbuild/buildtools#302]
Particularly when combined with issue #301 this makes it currently unusable for me, e.g.:

```
def _flags(values, defaults):
  return [flag
          for flag in [_flag(name, type(default), values.pop(name, default))
                       for name, default in defaults.items()]
          if flag != None]
```
While there are several reasonable, readable, formats for the above, the 139-character-single-line format buildifier chooses is none of them:

```
def _flags(values, defaults):
    return [flag for flag in [_flag(name, type(default), values.pop(name, default)) for name, default in defaults.items()] if flag != None]
```

**Comment by laurentlb** [comment]
It's not a question of "length limit", the issue is that the list comprehension should kept be on multiple lines.

This used to work in BUILD files.

**Comment by vladmos** [comment]
This used to work in BUILD files only if there was a line break between `[` and the first expression.

Do we need this issue or should I close it as a duplicate of #301? We can keep this one for fixing long lines, but once #301 is fixed buildifier will format the example above in the multiline mode.

**Comment by shahms** [comment]
I'm fine with dup'ing this to #301, I only split them up because they seemed like distinct problems.

**Comment by vladmos** [comment]
Maybe I don't get it, why are these problems distinct? The examples provided are the same, except in this one there are no suffix comments.

**Comment by shahms** [comment]
It sounds like they may not be distinct, but focusing on the similarity of the examples isn't useful.  It's easy for the same code snippet to demonstrate multiple formatting bugs and it's not always obvious which parts are related.  It's also easier to use the triggering example than it is to find a minimal reproduction of each issue.

"Unreadably long list comprehension" and "Drops EOL comments" seemed like distinct issues, both of which happened to present in a single example.

**Comment by vladmos** [comment]
The underlying issue is the same, buildifier shouldn't have formatted it in a single line, and in that case all suffix comments from the other example would have been printed on the correct lines.

If you want to have an issue about enforcing line length limits it's better create a new one because the example from this issue will be fixed with #309. Note that if you manually type such a long comprehension in one line it will remain multiline, but if you introduce at least one linebreak it will be multiline (the pull request contains the example with the fixed output in a test file).
